### PR TITLE
[LLAPI] Add RediSearch_IndexInfo

### PR DESCRIPTION
**Pull request by ashtul**

### CODE REVIEW

File: src/redisearch_api.h [hunk 1]
```diff
@@ -90,6 +91,61 @@ struct RSIdxOptions {
...
 
+struct RSIdxField {
+  char *path;
+  char *name;
+  struct {
+    bool text;
+    bool tag;
+    bool numeric;
+    bool geo;
+  } types;
```
Comment (AviAvni) [inline src/redisearch_api.h:102]: why not enum?

Comment (ashtul) [reply src/redisearch_api.h:102]: Because a field can have more than one type.

File: src/redisearch_api.c [hunk 2]
```diff
@@ -631,3 +638,91 @@ void RediSearch_SetCriteriaTesterThreshold(size_t num) {
...
+  }
+  if (specField->types == INDEXFLD_T_NUMERIC) {
+    infoField->types.numeric = true;
+  }
+  if (specField->types == INDEXFLD_T_TAG) {
+    infoField->types.tag = true;
+    infoField->tagSeperator = specField->tagSep;
+    infoField->tagCaseSensitive = specField->tagFlags & TagField_CaseSensitive ? 1 : 0;
+  }
+  if (specField->types == INDEXFLD_T_GEO) {
```
Comment (rafie) [inline src/redisearch_api.c:657]: shouldn't we be checking those doggies bitwise?

File: src/redisearch_api.h (hunk 1 above)
Comment (rafie) [reply src/redisearch_api.h:102]: Why can't this be a mask (i.e. `FieldType`)?

File: src/redisearch_api.c [hunk 3]
```diff
@@ -631,3 +638,91 @@ void RediSearch_SetCriteriaTesterThreshold(size_t num) {
...
+    // LLAPI always uses ForkGC
+    ForkGCStats gcStats = ((ForkGC *)sp->gc->gcCtx)->stats;
+
+    info->totalCollected = gcStats.totalCollected;
+    info->numCycles = gcStats.numCycles;
+    info->totalMSRun = gcStats.totalMSRun;
+    info->lastRunTimeMs = gcStats.lastRunTimeMs;
+  }
+
+  info->stopwords = RediSearch_IndexGetStopwords(sp, &info->stopwordsLen);
```
Comment (MeirShpilraien) [inline src/redisearch_api.c:715]: If we already have an api to get stopwords why we need it also here?

File: src/redisearch_api.c (hunk 2 above)
Comment (MeirShpilraien) [reply src/redisearch_api.c:657]: I agree, flags seems more suitable here.

File: src/redisearch_api.c [hunk 4]
```diff
@@ -631,3 +638,91 @@ void RediSearch_SetCriteriaTesterThreshold(size_t num) {
...
+  }
+  if (FieldSpec_IsNoStem(specField)) {
+    infoField->noStem = true;
+  }
+  if (FieldSpec_IsPhonetics(specField)) {
+    infoField->phonetic = true;
+  }
+  if (!FieldSpec_IsIndexable(specField)) {
+    infoField->noIndex = true;
+  }
```
Comment (MeirShpilraien) [inline src/redisearch_api.c:672]: Also those could be flags no?

File: src/redisearch_api.c [hunk 5]
```diff
@@ -631,3 +638,91 @@ void RediSearch_SetCriteriaTesterThreshold(size_t num) {
...
+  info->numFields = sp->numFields;
+  info->fields = rm_calloc(sp->numFields, sizeof(*info->fields));
+  for (int i = 0; i < info->numFields; ++i) {
+    RediSearch_FieldInfo(&info->fields[i], &sp->fields[i]);
+  }
+
+  info->numDocuments = sp->stats.numDocuments;
+  info->maxDocId = sp->docs.maxDocId;
+  info->docTableSize = sp->docs.memsize;
+  info->sortablesSize = sp->docs.sortablesSize;
```
Comment (MeirShpilraien) [inline src/redisearch_api.c:692]: Do we have sortable of LLAPI? I do not remember...

File: src/redisearch_api.c [hunk 6]
```diff
@@ -631,3 +638,91 @@ void RediSearch_SetCriteriaTesterThreshold(size_t num) {
...
+  info->skipIndexesSize = sp->stats.skipIndexesSize;
+  info->scoreIndexesSize = sp->stats.scoreIndexesSize;
+  info->offsetVecsSize = sp->stats.offsetVecsSize;
+  info->offsetVecRecords = sp->stats.offsetVecRecords;
+  info->termsSize = sp->stats.termsSize;
+  info->indexingFailures = sp->stats.indexingFailures;
+
+  if (sp->gc) {
+    // LLAPI always uses ForkGC
+    ForkGCStats gcStats = ((ForkGC *)sp->gc->gcCtx)->stats;
```
Comment (MeirShpilraien) [inline src/redisearch_api.c:707]: Wonder if we need to catch any lock to read those? Maybe this think should take the RWLock for read?

File: src/redisearch_api.c (hunk 3 above)
Comment (ashtul) [reply src/redisearch_api.c:715]: We print it with `FT.INFO`. Let's remove.

File: src/redisearch_api.c (hunk 5 above)
Comment (ashtul) [reply src/redisearch_api.c:692]: We have `RSFLDOPT_SORTABLE`.

File: src/redisearch_api.c (hunk 6 above)
Comment (ashtul) [reply src/redisearch_api.c:707]: `spec->stats` is updated as well in forkgc. They are updated on the parent with the lock taken.
@AviAvni is this command being called with the lock taken?

File: src/redisearch_api.h [hunk 7]
```diff
@@ -90,6 +91,61 @@ struct RSIdxOptions {
   RSLanguage lang;
 };
 
+struct RSIdxField {
+  char *path;
```
Comment (MeirShpilraien) [inline src/redisearch_api.h:95]: Wonder if we should version the struct for future changes and backward comparability

**Review by MeirShpilraien — COMMENTED** [review]
Looks good, some comment.

File: src/redisearch_api.h (hunk 1 above)
Comment (ashtul) [reply src/redisearch_api.h:102]: FieldType is OPAC for the LLAPI. Changing to use `RSFLDTYPE_`

File: src/redisearch_api.h (hunk 7 above)
Comment (ashtul) [reply src/redisearch_api.h:95]: Isn't that a concern when loading/saving to rdb?

File: src/redisearch_api.c (hunk 6 above)
Comment (AviAvni) [reply src/redisearch_api.c:707]: what lock? we don't lock redisearch explicit

Comment (MeirShpilraien) [reply src/redisearch_api.c:707]: @ashtul we have a RWLock internally insdie the spec. Look at other API's how they they the RWLock.

Comment (ashtul) [reply src/redisearch_api.c:708]: @MeirShpilraien  Added `RWLOCK_ACQUIRE_READ` and `RWLOCK_RELEASE`

Comment (ashtul) [reply src/redisearch_api.c:715]: Added `RS_INFO_CURRENT` and `RS_INFO_INIT` as version flags

File: src/redisearch_api.h (hunk 7 above)
Comment (ashtul) [reply src/redisearch_api.h:94]: Added `RS_INFO_CURRENT` and `RS_INFO_INIT` as version flags for support in future versions.

File: src/redisearch_api.c [hunk 8]
```diff
@@ -65,6 +65,13 @@ char **RediSearch_IndexGetStopwords(IndexSpec* sp, size_t *size) {
   return GetStopWordsList(sp->stopwords, size);
 }
 
+void RediSearch_StopwordsList_Free(char **list, size_t size) {
```
Comment (MeirShpilraien) [inline src/redisearch_api.c:68]: We did not have this API before?

**Review by MeirShpilraien — DISMISSED** [review]
Small question, not a blocker.

File: src/redisearch_api.c (hunk 8 above)
Comment (ashtul) [reply src/redisearch_api.c:68]: No

**Review by MeirShpilraien — APPROVED** [review]

**Pull request merged**